arroyo-api/src/rest.rs

```diff
@@ -1,32 +1,96 @@
 use utoipa::OpenApi;
 use utoipa_swagger_ui::SwaggerUi;
 
-#[derive(OpenApi)]
-#[openapi(
-    info(title = "Arroyo REST API", version = "1.0.0"),
-    paths(ping),
-    tags(
-        (name = "ping", description = "Ping")
-    )
-)]
-pub struct ApiDoc;
+use crate::pipelines::{
+    delete_pipeline, get_jobs, get_pipeline, get_pipelines, patch_pipeline, post_pipeline,
+};
+use crate::ApiDoc;
+use crate::ApiServer;
+use arroyo_types::{telemetry_enabled, API_ENDPOINT_ENV, ASSET_DIR_ENV};
+
+#[derive(Clone)]
+pub struct AppState {
+    pub(crate) grpc_api_server: ApiServer,
+    pub(crate) pool: Pool,
+}
 
 #[utoipa::path(
     get,
     path = "/v1/ping",
     tag = "ping",
     responses(
-        (status = 200, description = "Ping endpoint",),
+        (status = 200, description = "Ping endpoint"),
     ),
 )]
 pub async fn ping() -> impl IntoResponse {
     Json("Pong")
 }
 
-pub fn create_rest_app() -> Router {
+pub fn create_rest_app(server: ApiServer, pool: Pool) -> Router {
+    let asset_dir = env::var(ASSET_DIR_ENV).unwrap_or_else(|_| "arroyo-console/dist".to_string());
+
+    static INDEX_HTML: Lazy<String> = Lazy::new(|| {
+        let asset_dir =
+            env::var(ASSET_DIR_ENV).unwrap_or_else(|_| "arroyo-console/dist".to_string());
+
+        let endpoint = env::var(API_ENDPOINT_ENV).unwrap_or_else(|_| String::new());
+
+        std::fs::read_to_string(PathBuf::from_str(&asset_dir).unwrap()
+            .join("index.html"))
+            .expect("Could not find index.html in asset dir (you may need to build the console sources)")
+            .replace("{{API_ENDPOINT}}", &endpoint)
+            .replace("{{CLUSTER_ID}}", &arroyo_server_common::get_cluster_id())
+            .replace("{{DISABLE_TELEMETRY}}", if telemetry_enabled() { "false" } else { "true" })
+    });
+
+    let fallback = service_fn(|_: http::Request<_>| async move {
+        let body = Body::from(INDEX_HTML.as_str());
+        let res = http::Response::new(body);
+        Ok::<_, _>(res)
+    });
+
+    let serve_dir = ServeDir::new(&asset_dir).not_found_service(fallback);
+
+    // TODO: enable in development only!!!
+    let cors = CorsLayer::new()
+        .allow_headers(cors::Any)
+        .allow_origin(cors::Any);
+
+    let api_routes = Router::new()
+        .route("/ping", get(ping))
+        .route("/pipelines", post(post_pipeline))
+        .route("/pipelines", get(get_pipelines))
+        .route("/pipelines/:id", patch(patch_pipeline))
+        .route("/pipelines/:id", get(get_pipeline))
+        .route("/pipelines/:id", delete(delete_pipeline))
+        .route("/pipelines/:id/jobs", get(get_jobs));
+
     Router::new()
-        .merge(SwaggerUi::new("/swagger-ui").url("/api-docs/openapi.json", ApiDoc::openapi()))
-        .route("/v1/ping", get(ping))
+        .merge(
+            SwaggerUi::new("/api/v1/swagger-ui")
+                .url("/api/v1/api-docs/openapi.json", ApiDoc::openapi()),
+        )
+        .nest("/api/v1", api_routes)
```

Comment:
As discussed, `/api` 404s should fallback to a JSON error message instead of the HTML

Reply:
Done.